File: Astro-TFG/public/scripts/3-carga-graficos-complejos/prueba-1/version-1/javascript/main.js
let worker = null;
let workerTime = 0;

function initializeWorker() {
    if (!worker) {
        worker = new Worker(new URL("worker.js", import.meta.url), {
            type: "module",
        });
    }
}

async function _runJSBenchmark() {
    try {
        window.clearCell("javascript-output");

        let start_time_worker = performance.now();
        initializeWorker();
        workerTime = performance.now() - start_time_worker;

        const num_executions = parseInt(
            document.getElementById("num-executions-javascript").value
        ) || 1;

        const id = `js-${Date.now()}`;

        const resultJson = await new Promise((resolve, reject) => {
            function onMessage(e) {
                if (e.data.id !== id) return;
                worker.removeEventListener("message", onMessage);
                if (e.data.error) {
                    reject(new Error(e.data.error));
                } else if (e.data.json !== undefined) {
                    resolve(e.data.json);
                } else {
                    reject(new Error("Worker dont return results"));
                }
            }
            worker.addEventListener("message", onMessage);
            worker.postMessage({
                id,
                type: "do_analisis",
                size: 100_000,
                num_executions,
            });
        });

        const result = JSON.parse(resultJson);
        displayResult(result);
    } catch (error) {
        console.error("Worker error:", error);
    }
}


function createDiv() {
    const div = document.createElement("div");
    return div;
}

function displayResult(r) {
    const output = document.getElementById("javascript-output");

    const workerDiv = createDiv();
    workerDiv.textContent = `Worker init time: ${workerTime.toFixed(2)} ms`;
    output.appendChild(workerDiv);

    const avgDataGenDiv = createDiv();
    avgDataGenDiv.textContent = `Avg Data generation: ${r.data_gen_time.toFixed(2)} ms`;
    output.appendChild(avgDataGenDiv);

    const avgRenderDiv = createDiv();
    avgRenderDiv.textContent = `Avg Rendering: ${r.render_time.toFixed(2)} ms`;
    output.appendChild(avgRenderDiv);

    const avgExecutionTimeDiv = createDiv();
    avgExecutionTimeDiv.textContent = `Avg execution time: ${r.average_time_ms.toFixed(2)} ms`;
    output.appendChild(avgExecutionTimeDiv);

    const totalTimeDiv = createDiv();
    totalTimeDiv.textContent = `Total ET: ${r.total_time_ms.toFixed(2)} ms`;
    output.appendChild(totalTimeDiv);
    window.hideExecutionLoader();
    window.displayPlot(r.image_base64, "javascript-output");

}


window.runJsBenchmark = async function () {
    window.clearCell("javascript-output");
    window.showExecutionLoader();

    await new Promise(requestAnimationFrame);

    try {
        await _runJSBenchmark();
    } finally {
        window.hideExecutionLoader();
    }
};

File: Astro-TFG/public/scripts/3-carga-graficos-complejos/prueba-1/version-1/python/main.py
import asyncio
import time
import json
from pyscript import display, PyWorker
import js  # type: ignore

worker = None


async def initialize_worker():
    global worker
    if worker is None:
        base = js.window.location.origin + js.window.document.body.dataset.pyPath
        worker = PyWorker(
            f"{base}worker.py",
            type="pyodide",
            config=f"{base.replace('python', 'json')}pyscript-worker.json"
        )
        await worker.ready


async def launch_worker(event):
    global worker, worker_time
    js.clearCell('pyscript-output')

    start_w = time.perf_counter()
    await initialize_worker()
    worker_time = (time.perf_counter() - start_w) * 1000

    result_json = await worker.sync.do_graph_rendering()
    result = json.loads(result_json)

    display_result(result)


def display_result(result):
    global worker_time
    display(
        f"Worker init time: {worker_time:.2f} ms", target="pyscript-output")
    display(
        f"Avg Data generation: {result['data_gen_time']:.2f} ms", target="pyscript-output")
    display(
        f"Avg Rendering: {result['render_time']:.2f} ms", target="pyscript-output")
    display(
        f"Avg execution time: {result['average_time_ms']:.2f} ms", target="pyscript-output")
    display(
        f"Total ET: {result['total_time_ms']:.2f} ms", target="pyscript-output")
    if result["image_base64"]:
        js.window.displayPlot(
            result["image_base64"], "pyscript-output")
    js.window.hideExecutionLoader()


def js_run_py_benchmark(event):
    js.clearCell('pyscript-output')
    asyncio.ensure_future(launch_worker(None))


js.window.run_py_benchmark = js_run_py_benchmark
